text/0037-guarded-arith.md

```diff
@@ -0,0 +1,82 @@
+# Guard Arithmetic Expressions within Policies
+
+## Related issues and PRs
+
+- Reference Issues: No existing issue, but supersedes https://github.com/cedar-policy/rfcs/pull/36
+- Implementation PR(s): (Existing PR for range analysis based on schema: https://github.com/cedar-policy/cedar/pull/56)
+
+## Timeline
+
+- Started: 2023-11-07
+
+## Summary
+
+We will use a range analysis (as in https://github.com/cedar-policy/cedar/pull/56) but with the bounds given in the policy rather than in the schema.
+
+## Motivation
+
+Arithmetic with bounded integers is a common source of bugs [https://blog.research.google/2006/06/extra-extra-read-all-about-it-nearly.html]. Consider the following policy:
+
+```
+forbid(principal, action, resource) when {
+   principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms ||
+   !principal.is_mfa
+};
+```
+
+A casual reader of this policy may conclude that all actions are forbiden when `principal.mfa` is `false`. This is wrong. The lhs (`principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms`) is evaluated first and if it errors the entire policy is skipped.
+
+To avoid this class of mistakes, we want to change Cedar's validator to error on arithmetic expressions that could overflow. To pass validation, arithmetic should be guarded. For example, the above policy should be written:
+
+```
+forbid(principal, action, resource) when {
+   (principal.elevated_privileges_end_time_s < i64::MAX / 1000 &&
+    principal.elevated_privileges_end_time_s > i64::MIN / 1000 &&
+    principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms) ||
+   !principal.is_mfa
+};
+```
+
+## Detailed design
+
+Every non-identity preserving arithmetic operation should be guarded by comparisons that prevent over/underflow.
+
+- We add intervals to represent the ranges of each integer.
+- Integers without explicit intervals are assumed to lie in `[i64::MIN, i64::MAX]` so almost any arithmetic operation on them will be a validation error.
+- We use interval arithmetic, which is imprecise, but because the bounds are written in the policy, the intervals can be precisely given at the point of each operation.
+
+
+## Drawbacks
+
+The policies that use arithmetic and will pass validation are ugly.
+
+We assume almost no one uses arithmetic, so we think this is an acceptable tradeoff. If there are significant uses of arithmetic, a better approach is likely desirable.
+
+## Alternatives
+
+### Unbounded Integers
+Unclear how these will affect our FFI. We can take this proposal now and switch to unbounded ints later more easily than we can switch now and reverse that decision later.
+
+### Bounds in the schema.
+It is not clear how to analyze a set with members of an EntityType that has an integer attribute with bounds while keeping analysis sound, complete and decidable.
```

Comment:
It's worth elaborating on this point -- can you show an example? Would the issue only arise once/if `.any` and `.all` are in the language?

Reply:
This is an issue no matter what.  For example, without being able to enforce these  bounds, the solver could return false positives for set membership constraints, e.g., `principal.someNumbersLessThanTen.contains(1000)`.